### PR TITLE
Add single and multi branch cicd for awscodebuild

### PR DESCRIPTION
**Pull request by pohek321**

### CODE REVIEW

File: astro/ci-cd.md [hunk 1]
```diff
@@ -390,6 +390,98 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+    - `PROD_ASTRONOMER_KEY_ID`: Your Production Deployment API key ID
+    - `PROD_ASTRONOMER_KEY_SECRET`: Your Production Deployment API key secret
+    - `PROD_DEPLOYMENT_ID`: The Deployment ID of your Production Deployment
+    - `DEV_ASTRONOMER_KEY_ID`: Your Development Deployment API key ID
+    - `DEV_ASTRONOMER_KEY_SECRET`: Your Development Deployment API key secret
+    - `DEV_DEPLOYMENT_ID`: The Deployment ID of your Development Deployment
+
+2. At the root of your Git repository, add a [buildspec.yml](https://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html#build-spec-ref-example) that includes the following script:
+
+   <pre><code parentName="pre">{`
```
Comment (jwitz) [inline astro/ci-cd.md:452]:
```suggestion
   ```yaml
```

File: astro/ci-cd.md [hunk 2]
```diff
@@ -390,6 +390,98 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+           fi
+         - |
+           if expr "${CODEBUILD_WEBHOOK_HEAD_REF}" : "refs/heads/dev" >/dev/null; then
+             export ASTRONOMER_KEY_ID="${DEV_ASTRONOMER_KEY_ID}"
+             export ASTRONOMER_KEY_SECRET="${DEV_ASTRONOMER_KEY_SECRET}"
+             curl -sSL install.astronomer.io | sudo bash -s
+             astro deploy "${DEV_DEPLOYMENT_ID}" -f
+           fi
+   
+   `}</code></pre>
```
Comment (jwitz) [inline astro/ci-cd.md:478]:
```suggestion
   ```
```

File: astro/ci-cd.md [hunk 3]
```diff
@@ -390,6 +390,98 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+   
+     build:
+       commands:
+         - echo "${CODEBUILD_WEBHOOK_HEAD_REF}"
+         - export ASTRONOMER_KEY_ID="${ASTRONOMER_KEY_ID}"
+         - export ASTRONOMER_KEY_SECRET="${ASTRONOMER_KEY_SECRET}"
+         - curl -sSL install.astronomer.io | sudo bash -s
+         - astro deploy "${ASTRONOMER_DEPLOYMENT_ID}" -f
+   
+   `}</code></pre>
```
Comment (jwitz) [inline astro/ci-cd.md:432]:
```suggestion
   ```
```

File: astro/ci-cd.md [hunk 4]
```diff
@@ -390,6 +390,98 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+
+    - `ASTRONOMER_KEY_ID`: Your Deployment API key ID
+    - `ASTRONOMER_KEY_SECRET`: Your Deployment API key secret
+    - `ASTRONOMER_DEPLOYMENT_ID`: The Deployment ID of your production deployment
+
+    Be sure to set the values for your API credentials as secret.
+
+2. At the root of your Git repository, add a [buildspec.yml](https://docs.aws.amazon.com/codebuild/latest/userguide/build-spec-ref.html#build-spec-ref-example) file that includes the following script:
+
+   <pre><code parentName="pre">{`
```
Comment (jwitz) [inline astro/ci-cd.md:415]:
```suggestion
   ```
```

File: astro/ci-cd.md (hunk 3 above)
Comment (jwitz) [reply astro/ci-cd.md:432]: This suggestion _should_ read as adding three tick marks (`). Thanks GitHub!

File: astro/ci-cd.md (hunk 2 above)
Comment (jwitz) [reply astro/ci-cd.md:478]: This suggestion _should_ read as adding three tick marks (`). Thanks GitHub!

File: astro/ci-cd.md [hunk 5]
```diff
@@ -390,6 +390,96 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+             astro deploy "${PROD_DEPLOYMENT_ID}" -f
+           fi
+         - |
+           if expr "${CODEBUILD_WEBHOOK_HEAD_REF}" : "refs/heads/dev" >/dev/null; then
+             export ASTRONOMER_KEY_ID="${DEV_ASTRONOMER_KEY_ID}"
+             export ASTRONOMER_KEY_SECRET="${DEV_ASTRONOMER_KEY_SECRET}"
+             curl -sSL install.astronomer.io | sudo bash -s
+             astro deploy "${DEV_DEPLOYMENT_ID}" -f
+           fi
+
```
Comment (jwitz) [inline astro/ci-cd.md:476]:
```suggestion
    ```
```

File: astro/ci-cd.md [hunk 6]
```diff
@@ -390,6 +390,96 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+   
+     build:
+       commands:
+         - echo "${CODEBUILD_WEBHOOK_HEAD_REF}"
+         - export ASTRONOMER_KEY_ID="${ASTRONOMER_KEY_ID}"
+         - export ASTRONOMER_KEY_SECRET="${ASTRONOMER_KEY_SECRET}"
+         - curl -sSL install.astronomer.io | sudo bash -s
+         - astro deploy "${ASTRONOMER_DEPLOYMENT_ID}" -f
+   
+
```
Comment (jwitz) [inline astro/ci-cd.md:432]:
```suggestion
    ```
```

File: astro/ci-cd.md [hunk 7]
```diff
@@ -390,6 +390,96 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+       commands:
+         - echo "${CODEBUILD_WEBHOOK_HEAD_REF}"
+         - export ASTRONOMER_KEY_ID="${ASTRONOMER_KEY_ID}"
+         - export ASTRONOMER_KEY_SECRET="${ASTRONOMER_KEY_SECRET}"
+         - curl -sSL install.astronomer.io | sudo bash -s
+         - astro deploy "${ASTRONOMER_DEPLOYMENT_ID}" -f
+   
+    ```
+    When configuring the AWS CodeBuild project, create a webhook source event using the source provider where your Astro project is hosted (i.e. Bitbucket, GitHub, etc.). After doing so, configure the Primary source webhook events to trigger for event type `PUSH`. After setting those configurations, this buildspec.yml triggers a code push to an Astro Deployment every time a commit or pull request is merged to the `main` branch of your repository.
+</TabItem>
```
Comment (jwitz) [inline astro/ci-cd.md:434]:
```suggestion
    ```

    When configuring the AWS CodeBuild project, create a webhook source event using the source provider where your Astro project is hosted (i.e. Bitbucket, GitHub, etc.). After doing so, configure the Primary source webhook events to trigger for event type `PUSH`. After setting those configurations, this buildspec.yml triggers a code push to an Astro Deployment every time a commit or pull request is merged to the `main` branch of your repository.

</TabItem>
```

File: astro/ci-cd.md [hunk 8]
```diff
@@ -390,6 +390,98 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+       commands:
+         - echo "${CODEBUILD_WEBHOOK_HEAD_REF}"
+         - export ASTRONOMER_KEY_ID="${ASTRONOMER_KEY_ID}"
+         - export ASTRONOMER_KEY_SECRET="${ASTRONOMER_KEY_SECRET}"
+         - curl -sSL install.astronomer.io | sudo bash -s
+         - astro deploy "${ASTRONOMER_DEPLOYMENT_ID}" -f
+   
+    ```
+
+    When configuring the AWS CodeBuild project, create a webhook source event using the source provider where your Astro project is hosted (i.e. Bitbucket, GitHub, etc.). After doing so, configure the Primary source webhook events to trigger for event type `PUSH`. After setting those configurations, this buildspec.yml triggers a code push to an Astro Deployment every time a commit or pull request is merged to the `main` branch of your repository.
```
Comment (jwitz) [inline astro/ci-cd.md:434]:
```suggestion
3. In your AWS CodeBuild project, create a [webhook event](https://docs.aws.amazon.com/codebuild/latest/userguide/webhooks.html) for the source provider where your Astro project is hosted, such as BitBucket or GitHub. When configuring the webhook, select an event type of `PUSH`.

Your buildspec.yml file now triggers a code push to an Astro Deployment every time a commit or pull request is merged to the `main` branch of your repository.
```
A few questions here:
- What is the "source provider"? Is that where our Git repo is hosted?
- I found a "Primary source webhook" for GitHub but not for BitBucket. Can we generalize these instructions?

Even better, can we include an example configuration that someone could run using the [AWS CLI](https://docs.aws.amazon.com/cli/latest/reference/codebuild/create-webhook.html)?

File: astro/ci-cd.md [hunk 9]
```diff
@@ -390,6 +390,98 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+    ]}>
+<TabItem value="awscodebuildstandard">
+
+To automate code deploys to a single Deployment using [AWS CodeBuild](https://aws.amazon.com/codebuild/), complete the following setup in a Git-based repository hosting an Astro project:
+
+1. In your AWS CodeBuild pipeline configuration, add the following environment variables:
+
+    - `ASTRONOMER_KEY_ID`: Your Deployment API key ID
+    - `ASTRONOMER_KEY_SECRET`: Your Deployment API key secret
+    - `ASTRONOMER_DEPLOYMENT_ID`: The Deployment ID of your production deployment
```
Comment (jwitz) [inline astro/ci-cd.md:409]: We need to specify `CODEBUILD_WEBHOOK_HEAD_REF` here and explain what it does.

File: astro/ci-cd.md [hunk 10]
```diff
@@ -390,6 +390,98 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+
+<Tabs
+    defaultValue="awscodebuildstandard"
+    values={[
+        {label: 'Standard', value: 'awscodebuildstandard'},
+        {label: 'Multi-branch', value: 'awscodebuildmultibranch'},
+    ]}>
+<TabItem value="awscodebuildstandard">
+
+To automate code deploys to a single Deployment using [AWS CodeBuild](https://aws.amazon.com/codebuild/), complete the following setup in a Git-based repository hosting an Astro project:
```
Comment (jwitz) [inline astro/ci-cd.md:403]: I'm unfamiliar with CodeBuild, is your code stored in AWS or is it just a build tool? How is the buildspec triggered from GitHub/BitBucket/etc?

File: astro/ci-cd.md [hunk 11]
```diff
@@ -390,6 +390,98 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+
+</TabItem>
+
+<TabItem value="awscodebuildmultibranch">
+
+To automate code deploys across multiple Deployments using [AWS CodeBuild](https://aws.amazon.com/codebuild/), complete the following setup in a Git-based repository hosting an Astro project:
+
+1. In your AWS CodeBuild pipeline configuration, add the following environment variables:
+
+    - `PROD_ASTRONOMER_KEY_ID`: Your Production Deployment API key ID
```
Comment (jwitz) [inline astro/ci-cd.md:444]: Same here, need to specify `CODEBUILD_WEBHOOK_HEAD_REF`

File: astro/ci-cd.md [hunk 12]
```diff
@@ -390,6 +390,98 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+         - |
+           if expr "${CODEBUILD_WEBHOOK_HEAD_REF}" : "refs/heads/dev" >/dev/null; then
+             export ASTRONOMER_KEY_ID="${DEV_ASTRONOMER_KEY_ID}"
+             export ASTRONOMER_KEY_SECRET="${DEV_ASTRONOMER_KEY_SECRET}"
+             curl -sSL install.astronomer.io | sudo bash -s
+             astro deploy "${DEV_DEPLOYMENT_ID}" -f
+           fi
+    ```
+
+   When configuring the AWS CodeBuild project, create a webhook source event using the source provider where your Astro project is hosted (i.e. Bitbucket, GitHub, etc.). After doing so, configure the Primary source webhook events to trigger for event type `PUSH`. After setting those configurations, this buildspec.yml triggers a code push to an Astro Deployment every time a commit or pull request is merged to the `dev` and `main` branches of your repository.
```
Comment (jwitz) [inline astro/ci-cd.md:480]: Same edits from the previous setup apply here.

File: astro/ci-cd.md [hunk 13]
```diff
@@ -390,6 +390,98 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+   
+   phases:
+     install:
+       runtime-versions:
+         python: latest
+   
+     build:
+       commands:
+         - |
+           if expr "${CODEBUILD_WEBHOOK_HEAD_REF}" : "refs/heads/main" >/dev/null; then
```
Comment (jwitz) [inline astro/ci-cd.md:465]:
```suggestion
           if expr "${CODEBUILD_WEBHOOK_HEAD_REF}" : "refs/heads/main" >/prod/null; then
```

Comment (jwitz) [reply astro/ci-cd.md:465]: Do we need a different webhook for prod/ dev?

File: astro/ci-cd.md [hunk 14]
```diff
@@ -390,6 +390,98 @@ To automate code deploys across multiple Deployments using [Jenkins](https://www
...
+   
+    ```
+
+    When configuring the AWS CodeBuild project, create a webhook source event using the source provider where your Astro project is hosted (i.e. Bitbucket, GitHub, etc.). After doing so, configure the Primary source webhook events to trigger for event type `PUSH`. After setting those configurations, this buildspec.yml triggers a code push to an Astro Deployment every time a commit or pull request is merged to the `main` branch of your repository.
+
+</TabItem>
+
+<TabItem value="awscodebuildmultibranch">
+
+To automate code deploys across multiple Deployments using [AWS CodeBuild](https://aws.amazon.com/codebuild/), complete the following setup in a Git-based repository hosting an Astro project:
```
Comment (jwitz) [inline astro/ci-cd.md:440]:
```suggestion
To automate code deploys across multiple Deployments using [AWS CodeBuild](https://aws.amazon.com/codebuild/), complete the following setup in a Git-based repository in CodeBuild hosting an Astro project:
```
Is this correct?

**Review by jwitz — CHANGES_REQUESTED** [review]
Thanks for writing this up @pohek321 ! I added some comments and questions. If you could give me a brief overview of how CodeBuild interacts with the Git repo, that would help me a lot with my understanding. Is the Git repo hosted in CodeBuild? And if not, how does it communicate with CodeBuild?

File: astro/ci-cd.md (hunk 9 above)
Comment (pohek321) [reply astro/ci-cd.md:409]: Hey @jwitz! The CODEBUILD_WEBHOOK_HEAD_REF is an environment variable that is made available by default in AWS CodeBuild. When you are using a webhook to trigger a CodeBuild job, this variable is set with the branch where the webhook action originated from. So you don't need to set it manually.

File: astro/ci-cd.md (hunk 10 above)
Comment (pohek321) [reply astro/ci-cd.md:403]: Hey @jwitz with AWS CodeBuild, your code can be stored in GitHub, Bitbucket, etc. AWS CodeBuild uses a [webhook event](https://docs.aws.amazon.com/codebuild/latest/userguide/webhooks.html) to recognize when pushes, merges, and other git actions happen in a repository to trigger the AWS CodeBuild pipeline. Jenkins behaves similarly to this as well.

File: astro/ci-cd.md (hunk 11 above)
Comment (pohek321) [reply astro/ci-cd.md:444]: Hey @jwitz! The CODEBUILD_WEBHOOK_HEAD_REF is an environment variable that is made available by default in AWS CodeBuild. When you are using a webhook to trigger a CodeBuild job, this variable is set with the branch where the webhook action originated from. So you don't need to set it manually.

File: astro/ci-cd.md (hunk 13 above)
Comment (pohek321) [reply astro/ci-cd.md:465]: @jwitz 

No need! I have tested this script in action and both need to be `/dev/null` because that's the result of a BASH expression when two strings are equal.

File: astro/ci-cd.md (hunk 14 above)
Comment (pohek321) [reply astro/ci-cd.md:440]: I would say that this isn't correct. CodeBuild doesn't behave as a Git repository, but a CI/CD tool. Some Git repository services contain built-in CI/CD tools (i.e. GitHub Actions, GitLab, BitBucket), but there are other standalone CI/CD tools out there that do not act as Git repositories (i.e. Jenkins, CircleCI, AWS CodeBuild, etc.)

File: astro/ci-cd.md (hunk 11 above)
Comment (pohek321) [reply astro/ci-cd.md:444]: So basically, it helps the CI/CD process identify if the developer pushed commits to dev/staging/main/etc. and we can build conditional logic (based on the branch they are working in) to trigger a deploy to an environment associated to that branch.

File: astro/ci-cd.md (hunk 8 above)
Comment (pohek321) [reply astro/ci-cd.md:434]: Correct, source provider is where the Git repo is hosted (i.e. GitHub, BitBucket)! I think BitBucket, GitHub, and AWS CodeCommit are the only 3 options for Git source repo's in AWS CodeBuild.

Are you asking for a configuration that someone could use to set up this same AWS CodeBuild process via the AWS CLI? I'm sure that the AWS CLI commands exist to set this up, but that wouldn't be uniform with the rest of the CI/CD docs on this page and what we recommend as a starting point (See GitHub Actions, CircleCI, Jenkins, etc.).

**Review by jwitz — APPROVED** [review]
Ahh thanks for clearing up my confusion here @pohek321, I hadn't realized that webhooks are configured initially from the source provider and can then be subscribed to from CodeBuild. Please review the two final copy suggestions here and commit them if you agree - after that, this looks great!
